utils/tests_fetcher.py

```diff
@@ -0,0 +1,387 @@
+
+    should_be_tested = set(not_touched_test_files) - set(EXPECTED_TEST_FILES_NEVER_TOUCHED)
+    if len(should_be_tested) > 0:
+        raise ValueError(
+            "The following test files are not currently associated with any module or utils files, which means they "
+            f"will never get run by the CI:\n{_print_list(should_be_tested)}\n. Make sure the names of these test "
+            "files match the name of the module or utils they are testing, or adapt the constant "
+            "`SPECIAL_MODULE_TO_TEST_MAP` in `utils/tests_fetcher.py` to add them. If your test file is triggered "
+            "separately and is not supposed to be run by the regular CI, add it to the "
+            "`EXPECTED_TEST_FILES_NEVER_TOUCHED` constant instead."
+        )
+
+
+def infer_tests_to_run(output_file):
+    modified_files = get_modified_python_files()
+    print(f"\n### MODIFIED FILES ###\n{_print_list(modified_files)}")
+
+    # Create the map that will give us all impacted modules.
+    impacted_modules_map = create_reverse_dependency_map()
+    impacted_files = modified_files.copy()
+    for f in modified_files:
+        if f in impacted_modules_map:
+            impacted_files.extend(impacted_modules_map[f])
+
+    # Remove duplicates
+    impacted_files = sorted(list(set(impacted_files)))
+    print(f"\n### IMPACTED FILES ###\n{_print_list(impacted_files)}")
+
+    # Grab the corresponding test files:
+    test_files_to_run = []
+    for f in impacted_files:
+        # Modified test files are always added
+        if f.startswith("tests/"):
+            test_files_to_run.append(f)
+        else:
+            new_tests = module_to_test_file(f)
+            if new_tests is not None:
+                if isinstance(new_tests, str):
+                    test_files_to_run.append(new_tests)
+                else:
+                    test_files_to_run.extend(new_tests)
+
+    # Remove duplicates
+    test_files_to_run = sorted(list(set(test_files_to_run)))
+    print(f"\n### TEST TO RUN ###\n{_print_list(test_files_to_run)}")
+    if len(test_files_to_run) > 0:
+        with open(output_file, "w", encoding="utf-8") as f:
+            f.write(" ".join(test_files_to_run))
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--sanity_check", action="store_true", help="Whether to just perform a sanity check.")
+    parser.add_argument(
+        "--output_file", type=str, default="test_list.txt", help="Where to store the list of tests to run"
+    )
+    args = parser.parse_args()
+    if args.sanity_check:
+        sanity_check()
+    else:
+        repo = Repo(PATH_TO_TRANFORMERS)
+        if not repo.head.is_detached and repo.head.ref == repo.refs.master:
+            print("Master branch detected, running all tests.")
+            with open(args.output_file, "w", encoding="utf-8") as f:
+                f.write("./tests/")
```

Comment:
I assume this will never hit the job that runs `examples`.

Reply:
No, the examples job just looks if the file exists or not, not its content.